realm/realm-library/src/main/java/io/realm/OrderedCollectionChangeSet.java

```diff
@@ -28,6 +30,64 @@
  * change, or an array of {@link Range}s.
  */
 public interface OrderedCollectionChangeSet {
+
+    /**
+     * State describing the nature of the changeset.
+     */
+    public enum State {
+        /**
+         * This state is used first time the callback is invoked. The query will have completed and
+         * data is ready for the UI.
+         */
+        INITIAL,
```

Comment:
So this presents the state where the query returns the first time and null change set passed to user before? 

Please notice there might be some error cases for the sync query -- Object Store don't distinguish the sync & async queries. So we had some hacks in the java to simulate our old sync query behaviour -- the first time callback called for sync query, the change set is not null or empty, it is supposed to contain the changeset between the data at the sync query created and the latest. Which means, to match that behavior, the state should be `UPDATE` here . So sync query doesn't have `INITIAL` state.

But it is actually quite a lot headache for us, see https://github.com/realm/realm-object-store/issues/605

I would suggest that we totally follow the cocoa API, merge `findAll` & `findAllAsync` into one API, all the queries are async by default (and they are actually already.). So every query callback will have an `INITIAL` state. Since we have a chance to make breaking changes, we'd better do it now to save lots of time in the future.

Reply:
Good point about synchronous queries not really having an `INITIAL` state, but only `UPDATES`.

But I'm not sure that following Cocoas API is actually the right choice, to be honest. I mean, I can see it will help _us_ on the internal side, but it is a quite severe breaking change that also causes a more unsafe and confusing API, as if you by accident do the wrong thing, then you switched an async query to be synchronous. This is what annoys me most about having a single API, nothing prevents you from shooting yourself in the foot, and Android developers have historically wanted more control over their threads than Cocoa devs.

With regard to naming  then:
If we combine them in `findAll()`: It breaks our naming pattern used everywhere else that suffixes `Async` to API's that uses a worker thread.

If we combine them in `findAllAsync()`: Then we have to explain to people that touching any property outside a change listener will cause it to evaluate the query immediately.

Part of this issue is also that the Java API returns a `RealmResult` immediately instead of a token, which makes it a lot easier to do wrong things by accident.

If we followed the Cocoa API, our API would look like this:

```
NotificationToken token = realm.where(Person.class).findAllAsync(new OrderedRealmCollectionChangeListener() {
    @Override
    public voic onChange(RealmResults<Person> persons, OrderedColllectionChangeSet changeSet) { 
     // ... 
    }
});
```

No Android/Java API I have ever seen does it like this.

Reply:
No need to expose the token to users. We still return the `RealmResults`.

If we can solve 
https://github.com/realm/realm-object-store/issues/605 , the current API is also good and easy to maintain.

Reply:
@cmelchior

    Disposable disposable = observable.subscribeWith((newValue) -> {
        //...
    });

Reply:
@Zhuinden You just had to prove me wrong 😉 

But that is actually a great example. Thanks 👍